LCLSGeneral/LCLSGeneral/POUs/Data/FB_BasicStats.TcPOU

```diff
@@ -90,19 +97,22 @@ IF bReset THEN
     fMax := 0;
     fMin := 0;
     fRange := 0;
+    fRMS := 0;
     bValid := FALSE;
     bReset := FALSE;
 ELSIF NOT (bExecute OR bAlwaysCalc) THEN
     bValid := FALSE;
 ELSIF bAlwaysCalc OR rTrig.Q THEN
-    // First pass through aSignal: get sum, mean, max, min
+    // First pass through aSignal: get sum, mean, max, min, rms
     nElemsSeen := 0;
     fSum := 0;
+    fRMSSum := 0;
     fMax := aSignal[LOWER_BOUND(aSignal, 1)];
     fMin := fMax;
     FOR nIndex := LOWER_BOUND(aSignal, 1) TO UPPER_BOUND(aSignal, 1) DO
         nElemsSeen := nElemsSeen + 1;
         fSum := fSum + aSignal[nIndex];
+        fRMSSum := fRMSSum + EXPT(aSignal[nIndex], 2);
```

Comment:
Double-check the RMS formula - I believe the average (mean) should make it in here somewhere.

Reply:
lgtm, RMS is the sum of the squares divided by the number of measurements then square rooted. See line 128. Will test this though in case I am missing something.

Reply:
checked out, RMS was 45 nrad, about what we expected, and almost equal to ~stddev

Reply:
I think this looks correct to me comparing to the wiki page: https://en.wikipedia.org/wiki/Root_mean_square
We can double-check its work if the IOC is live, because you can caget the full array and the calculated RMS and do the math yourself in e.g. python/numpy

Reply:
I'm starting to get confused though:
- In the example screenshot, the variance is higher than the rms. Is that possible?
- The below calculation for variance uses the mean

Complicating matters:
- RMS should be calculated with the raw encoder value as an input, right?
- The input for the optics PLC is essentialy the posdif in units of encoder counts, rather than the raw counts

Am I completely off-base here? My brain has been objectively confused today so I hope this isn't contagious

Reply:
- I don't see variance in the screenshot.
- RMS in this case is being calculated as a difference of ENC counts from the set point and is using current raw encoder counts to accomplish that. The one exception is the backward translation of set point from actual to encoder count.

Reply:
<img width="1007" alt="Screenshot 2023-10-11 at 2 56 46 PM" src="https://github.com/pcdshub/lcls-twincat-general/assets/79480290/0ff4ab3f-8fa3-49ce-937b-05e94a7bb6a0">

Reply:
Ugh, my brain is exploding today: I meant the standard deviation is higher than the rms

Reply:
I think I have a fundamental misunderstanding as to how to understand these two quantities.
The one thing I'd really love to do is to just "check the calculation" but I don't see an obvious way to do that

Reply:
I am not sure how to relate the standard dev to the rms of the error. On the one hand they both likely deviate about the same, but they ultimately calculate that difference from totally different places. One being the average and the other the set point. I would actually expect the latter to be higher, since the magnitude of deviation from a static point seems more likely to be higher than the deviation from an average.

Reply:
I took a break, got some fresh air, and then re-read the wiki articles.
I think this code calculates everything correctly and the size differences between the rms and stdev should definitely be small (since we're sending values whose mean are approximately zero). I think the RMS should always be slighly higher.

Formula pastes that make the most sense to me and match 1:1 with the code:
![image](https://github.com/pcdshub/lcls-twincat-general/assets/10647860/65b8b15c-75c5-448a-90ac-8351f070dbaa)
![image](https://github.com/pcdshub/lcls-twincat-general/assets/10647860/70fc59ff-400e-4350-b02a-6702490e648e)

Reply:
With respect to these quantities in the specific case of the mono encoder: the stdev is meaningful no matter how we input the encoder values, the rms is only meaningful if we offset them by e.g. the setpoint. If we offset the rms inputs by the mean we get the stdev, instead we're offseting them by the setpoint which is the more meaningful quantity for us.